### PR TITLE
[compiler-v2] Emitting missing FreezeRef in bytecode_generator

### LINKED ISSUE #12394: [Bug][Compiler-v2] STLOC_TYPE_MISMATCH_ERROR when the return type is a tuple containing mutable refs while the target type is immutable ref
# 🐛 Bug

When compiling the following code:

```
fun t2(u1: &mut u64, u2: &mut u64): (&mut u64, &mut u64) {
        (u1, u2)
    }
    fun test_4() {
         let x: &u64;
         let y: &u64;
         (x, y) = t2(&mut 3, &mut 4);
   }
```

The bytecode verifier will generate an error with the status code `STLOC_TYPE_MISMATCH_ERROR`.

### LINKED ISSUE #14274: [compiler-v2] Type Safety verification fails with STLOC_TYPE_MISMATCH_ERROR
Given the following code:

```
module 0x8675309::M {
    struct R has key { f: u64 }

    fun t4(cond: bool, addr: address): bool acquires R {
        let r = R { f: 0 };
        let f = &borrow_global<R>(addr).f;
        let r1; if (cond) r1 = borrow_global<R>(addr) else r1 = &mut r;
        let res = f == &r1.f;
        R { f: _ } = r;
        res
    }
}

```

... we get (with new borrow analysis, but unrelated to it):

```
Diagnostics:
bug: bytecode verification failed with unexpected status code `STLOC_TYPE_MISMATCH_ERROR`. This is a compiler bug, consider reporting it.
   ┌─ tests/reference-safety/v1-tests/borrow_global.move:29:27
   │
29 │         let r1; if (cond) r1 = borrow_global<R>(addr) else r1 = &mut r;
   │                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

### PR DESCRIPTION
**Pull request by wrwg**
## Description

FreezeRef was missing in assignments, including tuple assignment `(x, y) = f(..)`.
    
Fixes #14274
Fixes #12394

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Existing tests now pass

### CODE REVIEW

**Comment by wrwg** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14280?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14280** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14280?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14276** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14276?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @wrwg and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.move
Comment (brmataptos) [inline third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.move:1]: Do you want to remove the `TODO`?

**Review by brmataptos — APPROVED** [review]

File: third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.exp
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.exp:1]: Could you add a test case with initialization of tuple value: `let (x, y) = f(..)`?

**Review by rahxephon89 — APPROVED** [review]

File: third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.move [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    fun test_4() {
+        let _x: &u64;
+        let _y: &u64;
+        (_x, _y) = t2(&mut 3, &mut 4);
+    }
+
+    fun test_5() {
+        let (_x, _y): (&u64, &u64) = t2(&mut 3, &mut 4);
+    }
+}
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.move:14]: Could we please add the following (transactional) test case:

```
//# publish
module 0xc0ffee::m {
    public fun test(): u64 {
        let a = 1;
        let (x, y): (&u64, &u64) = (&mut a, freeze(&mut a));
        *x + *y
    }
}

//# run 0xc0ffee::m::test
```

On this branch, the above code gets a complier error in v1, but successfully executes in v2.

If you think this should be a bug filed, let me know and I can do so.

File: third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.move [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module 0x815::m {
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.move:1]: Another test to add here or as a new one (this works correctly btw):

```
    fun foo(x: &mut u64): &u64 {
        *x = 2;
        x
    }

    fun bar(x: &u64): &u64 {
        x
    }

    public fun test(x: u64): u64 {
        *bar(foo(&mut x))
    }
```

**Review by vineethk — APPROVED** [review]
Approving, but I have a failing test case suggested that may or may not be relevant to this PR.

File: third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.move (hunk 1 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/reference-safety/bug_12394_missing_freeze.move:14]: Added it, and that is what I'm seeing here too. Seems to me a bug in v1 borrow analysis, which wee aren't going to fix. I added a comment to clarify this in the test case.

**Pull request merged**